### ISSUE TITLE
keras layers LSTM uses inconsistent dropout approach

### ISSUE THREAD

**Issue by zacwellmer** [tensorflow/tensorflow#40215]
**System information**
- TensorFlow version (use command below): 2.2

**Describe the current behavior**
The input dropout mask is the same for the respective input, forget, update, and output computation. 

from the call method in the LSTM class in [recurrent_v2.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/layers/recurrent_v2.py)
```python
    dropout_mask = self.get_dropout_mask_for_cell(inputs, training, count=4)
    if dropout_mask is not None:
        inputs = inputs * dropout_mask[0]
```

However, the recurrent dropout mask is unique for the input, forget, update, and output computation
```python
        h_tm1_i = h_tm1 * rec_dp_mask[0]
        h_tm1_f = h_tm1 * rec_dp_mask[1]
        h_tm1_c = h_tm1 * rec_dp_mask[2]
        h_tm1_o = h_tm1 * rec_dp_mask[3]
```

If the input dropout mask approach was the intended behavior can someone link me a reference paper that explains why and shouldn't we at least set the count=1. To me it seems like we are attempting to follow the approach from "[A Theoretically Grounded Application of Dropout in Recurrent Neural Networks](http://papers.nips.cc/paper/6241-a-theoretically-grounded-application-of-dropout-in-recurrent-neural-networks.pdf)"

**Describe the expected behavior**
I'd have thought the input dropout mask would be unique to the input, forget, update, and output computation

**Comment by Saduf2019** [comment]
@zacwellmer 
Please share a simple stand alone code such that we can replicate the issue faced, or is possible share a colab gist for us to analyse the error faced.

**Comment by zacwellmer** [comment]
I don't think stand alone code would help to describe this issue since the issue is heavily embedded into the LSTM implementation. I'll write the equations which are being performed and what I expect to occur.

# Current Behavior
for each sequence the masks $\mathbf{m}\_{x}$, $\mathbf{m}\_{hi}$, $\mathbf{m}\_{hf}$, $\mathbf{m}\_{ha}$, and $\mathbf{m}\_{ho}$ are sampled then used in the following LSTM update
![image](https://user-images.githubusercontent.com/9603276/83980653-54ef8700-a8cc-11ea-9094-8404ac483da3.png)
![image](https://user-images.githubusercontent.com/9603276/83980476-11e0e400-a8cb-11ea-84b1-a33f5c9641d8.png)
![image](https://user-images.githubusercontent.com/9603276/83980483-245b1d80-a8cb-11ea-8ef9-c8c5a9792cd5.png)

# Expected Behavior
I would expect $\mathbf{m}\_{xi}$, $\mathbf{m}\_{xf}$, $\mathbf{m}\_{xa}$, and $\mathbf{m}\_{xo}$ to be different for each gate
![image](https://user-images.githubusercontent.com/9603276/83980529-83b92d80-a8cb-11ea-86ce-fbae1b700e4a.png)

I'm guessing this is not currently being done b/c there is a desire to use the gpu implementation and that doesn't seem to allow 4 different masked ```inputs``` values

**Comment by zacwellmer** [comment]
@qlzh727 is there any updates on this? If there's a paper that the current convention was modeled after I'd love to check it out. If not, I'm happy to help out with a PR, though it seems like this would change _could_use_gpu_kernel to False for input dropout. 

Would it be possible to just mask the kernel (could be done to have the same effect as masking the input for forward prop)? I think this would mean we need to mask gradients as well though. If this is possible, we could use the same solution for recurrent dropout to enable _could_use_gpu_kernel

**Comment by zacwellmer** [comment]
@qlzh727 @haozha111 do you folks have any thoughts on this? I'm open to make the PR to fix this but would prefer to know if it's a bad idea before spending time on it

**Comment by qlzh727** [comment]
Thanks for reporting the issue. Let me check the change history and related paper/math, and I will post some updates.

**Comment by zacwellmer** [comment]
@qlzh727 is there any updates on this? I'm still open to helping

**Comment by zacwellmer** [comment]
@qlzh727 FYI looks there is also discrepancies in how the input mask is handled between when _could_use_gpu_kernel is True and False. When it is True it looks as though we properly apply dropout (uniquely to i, f, a, and o) ([here](https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/layers/recurrent.py#L2388)).
```python
        inputs_i = inputs * dp_mask[0]
        inputs_f = inputs * dp_mask[1]
        inputs_c = inputs * dp_mask[2]
        inputs_o = inputs * dp_mask[3]
```

Seems like the simplest fix for this is to add an extra condition on ```self._could_use_gpu_kernel``` requiring that ```dropout==0.0```. Then the following lines could be dropped ([here](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/layers/recurrent_v2.py#L1146))
```
self.reset_dropout_mask()
      dropout_mask = self.get_dropout_mask_for_cell(inputs, training, count=4)
      if dropout_mask is not None:
        inputs = inputs * dropout_mask[0]
```

**Comment by qlzh727** [comment]
Very sorry for the long wait. 

I checked RNN code history in recurrent.py. Historically we have 2 implementation, which is branched by flag "implementation" in https://github.com/tensorflow/tensorflow/blob/8c80b0433f6f09a720a4fc2fbe55b64b4948965c/tensorflow/python/keras/layers/recurrent.py#L2429.

In the implementation 1, the input/recurrent input to IFCO gates are dropout via different mask, whereas in implementation 2, they are dropout via same mask. See https://github.com/tensorflow/tensorflow/blob/8c80b0433f6f09a720a4fc2fbe55b64b4948965c/tensorflow/python/keras/layers/recurrent.py#L2468. The recurrent_v2 is using implementation 2 as the basic and mimic it behavior. Historically, we create 4 dropout mask so that both implementation can use the same field, rather than branching the mask creation logic based on "implementation" version.

Hope this answers your question.

**Comment by zacwellmer** [comment]
No worries and thanks for circling back! 

I understand what it is doing but I don't understand why this would be the desired behavior. I don't understand why a different variant of input dropout would silently be used from an arbitrary combination of flags (ex: ```dropout>0``` and ```recurrent_dropout=0``` vs. ```dropout>0``` and ```recurrent_dropout>0``` ).

The only case I could see for sharing the IFCO dropout masks is if the user preferred the code to be faster (using the cudnn implementation) at the cost of questionable performance (I'm not aware of any works studying the impact of fixing a mask for all gates). But even still why would we only offer this speed/performance trade-off for the input mask but not the hidden state mask?

**Comment by qlzh727** [comment]
The reason we split the logic between dropout = 0 and recurrent_dropout = 0 in recurrent_v2 implementation is that cudnn kernel is a fused op and we can't inject the recurrent dropout for hidden state in each timestep. On the other hand, the input dropout can be done by dropping out the overall input sequence before feeding the input to the cudnn kernel.

Note that in the implementation 2, the behavior between recurrent.py and recurrent_v2.py should be the same. The same dropout mask is used for all 4 gates, and I think that's intended. We could certainly change recurrent_v2.py to use 4 different mask for input IFCO, but I am not sure if that's intended behavior. I would expect the mask to randomly drop the same feature for all IFCO, which mimic the behavior that some certain feature is totally hide from the network within the batch. This is aligned with the concept of the variational dropout, which use the same mask for all timesteps. 

It seems to me that using different mask for 4 gates will introduce noise to the network, but I don't have the math proof for it.

**Comment by zacwellmer** [comment]
ah gotcha, thanks for the explanation. I'm not sure how the cudnn implementation works but the reasoning sounds like it makes sense 

> I would expect the mask to randomly drop the same feature for all IFCO, which mimic the behavior that some certain feature is totally hide from the network within the batch. This is aligned with the concept of the variational dropout, which use the same mask for all timesteps.

This approach is called "tied-weights LSTM". I agree that the tied-weights approach seems to closely follow traditional uses of dropout (dropping a specific subset of units). However, dropout can also be interpreted as dropping rows of the weight matrix. Through the lens of the second approach it seems intuitive that the mask should differ between gates. 

It turns out the untied-weights LSTM performs better. From [section 4.1](http://papers.nips.cc/paper/6241-a-theoretically-grounded-application-of-dropout-in-recurrent-neural-networks.pdf) and referring to the trade-off from tied-weights

> With the second parametrisations we would place a distribution over the single matrix W. This leads to a faster forward-pass, but with slightly diminished results as we will see in the experiments section

From Table 1
![image](https://user-images.githubusercontent.com/9603276/92803715-89139180-f36c-11ea-87d3-64eecaa4ca16.png)

> It seems to me that using different mask for 4 gates will introduce noise to the network

yeah, the entropy of q(ω) would be larger in the un-tied weights approach

**Comment by qlzh727** [comment]
Thanks for the detailed reference. 

If I understand the "MC" correctly, I don't think keras is currently using MC dropout at test time. Basically dropout is disabled at eval/predict time. And if we remove the item with MC, then the tied weights are actually out performing the untied weights (lower error and higher WPS).

**Comment by zacwellmer** [comment]
that's a great point, thanks for taking a closer look at that table than I did. On second thought I'm not sure how helpful that table is for us since it doesn't tell us which performs better when dropout is only applied to the input gates (implementation=1).

Still seems odd to me that we follow the tied weights approach when ```dropout>0``` and ```recurrent_dropout=0``` but follow the untied approach when ```dropout>0``` and ```recurrent_dropout>0```. But if this is the desired behavior, I guess I could at least update the docs

**Comment by qlzh727** [comment]
I don't think we perform differently between recurrent_dropout=0 and recurrent_dropout>0, since the default implementation in TF2 is implementation=2, which only use the first mask out of 4 to mask the input. The implementation=1 is only used in TF1 and we can't change its behavior since it is suppose to be frozen.

https://github.com/tensorflow/tensorflow/blob/8c80b0433f6f09a720a4fc2fbe55b64b4948965c/tensorflow/python/keras/layers/recurrent.py#L2469

If there is no other questions, I will close this bug as working as intended. Feel free to reopen it if there is any other issue.

Sorry again for the long wait and thank you the bring up this question to us.

**Comment by zacwellmer** [comment]
> I don't think we perform differently between recurrent_dropout=0 and recurrent_dropout>0, since the default implementation in TF2 is implementation=2, which only use the first mask out of 4 to mask the input.

I don't think this is true. ```implementation=1``` is still used when ```dropout>0``` and ```recurrent_dropout=0```. I couldn't figure out a good way to fix the generated dropout masks (fixing the random seed doesn't work) between both approaches. You will need to follow this hacky approach with pdb. In short, the old approach is still used in [recurrent_v2.py](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/layers/recurrent_v2.py#L1149) 

Here's a [colab notebook](https://colab.research.google.com/drive/1o7DWrvkPZxiu2Ovww-XeCRGbWlsB-xhP?usp=sharing) to demonstrate
![image](https://user-images.githubusercontent.com/9603276/92985048-ed654c80-f463-11ea-9a26-5f5c4754ac85.png)

**Comment by qlzh727** [comment]
> > I don't think we perform differently between recurrent_dropout=0 and recurrent_dropout>0, since the default implementation in TF2 is implementation=2, which only use the first mask out of 4 to mask the input.
> 
> I don't think this is true. `implementation=1` is still used when `dropout>0` and `recurrent_dropout=0`. I couldn't figure out a good way to fix the generated dropout masks (fixing the random seed doesn't work) between both approaches. You will need to follow this hacky approach with pdb. In short, the old approach is still used in [recurrent_v2.py](https://github.com/tensorflow/tensorflow/blob/v2.2.0/tensorflow/python/keras/layers/recurrent_v2.py#L1149)
> 
> Here's a [colab notebook](https://colab.research.google.com/drive/1o7DWrvkPZxiu2Ovww-XeCRGbWlsB-xhP?usp=sharing) to demonstrate
> ![image](https://user-images.githubusercontent.com/9603276/92985048-ed654c80-f463-11ea-9a26-5f5c4754ac85.png)

I don't think so.. when dropout>0 and recurrent_dropout=0, it will fallback to recurrent.py, but both LSTM and LSTM Cell are created with implementation=2 (you can verify that by print out the layer.implementation). When implementation=2, it will use tied dropout, whereas implementation=1 will use untied dropout.

**Comment by zacwellmer** [comment]
Regardless of what the user passes input for ```implementation```, if ```self._could_use_gpu_kernel``` is true, then the tied weights approach is used for inputs.
https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/layers/recurrent_v2.py#L1091-L1095

https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/layers/recurrent_v2.py#L1120-L1149

> it will fallback to recurrent.py

Maybe I'm misunderstanding something. By "fallback to recurrent.py" do you mean that recurrent_v2.py is not used? b/c that example in the colab notebook shows the tied weights approach being used (unless the output is thrown-away after?)

**Comment by qlzh727** [comment]
Correct, when self._could_use_gpu_kernel is True, then the tied dropout approach is used and all the logic is in recurrent_v2.py. When self._could_use_gpu_kernel is False, it will fallback to https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/layers/recurrent_v2.py#L1120-L114, and the logic of self.cell.call() is the actual logic gets triggered which lives in recurrent.py. When implementation=2, then the tied dropout will be used in that case.

**Comment by zacwellmer** [comment]
The user provided input of the implementation flag doesn't seem to matter since it's overridden anytime the recurrent dropout rate is greater than 0. 
https://github.com/tensorflow/tensorflow/blob/2b96f3662bd776e277f86997659e61046b56c315/tensorflow/python/keras/layers/recurrent.py#L2300-L2302

The tied weights dropout looks like it is _only_ used when ```implementation=2```, ```dropout>0```, and ```recurrent_dropout=0```. 

Why wouldn’t we just offer a tied weights approach and an untied weights approach instead of silently changing based on an arbitrary set of parameters which the user is unaware of?
